### PR TITLE
Windows support [part 2]

### PR DESCRIPTION
**Pull request by petrutlucian94**
Windows support [part 2]

This PR series intends to set the ground for the upcoming Windows port effort.
For now, we're mostly interested in the client side, allowing Windows hosts
to consume rados, rbd and cephfs resources.

Those PRs should be reviewed and merged in order. Each PR depends on the previous one from the series (in fact it includes the previous commits as well).

~~Part 1: #31981~~
**Part 2**: #32704
Part 3: #32705
Part 4: #32707
Part 5: #32780 
Part 6: #32779
Part 7: #32778 
Part 8: #32777 
Part 9: #32776
Part 10: #33750

Related but independent PRs:
https://github.com/ceph/ceph/pull/32027
https://github.com/ceph/fmt/pull/2
https://github.com/ceph/rocksdb/pull/42

For more details about the build process and current status, please check the
README.windows.rst file. Worth mentioning that the resulted binaries can
connect to ceph clusters and consume pools: http://paste.openstack.org/raw/787534/

Here are some test results as well:
[test_results.zip](https://github.com/ceph/ceph/files/4077517/test_results.zip)

Any feedback is highly appreciated.

Signed-off-by: Lucian Petrut lpetrut@cloudbasesolutions.com
Signed-off-by: Alin Gabriel Serdean aserdean@cloudbasesolutions.com

### CODE REVIEW

**Comment by petrutlucian94** [comment]
jenkins test make check

**Comment by petrutlucian94** [comment]
jenkins test make check

File: src/common/shared_lib.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+int close_shared_lib(lib_handle handle);
+// Returns a string describing the most recent error.
+// Uses dlerror on Posix compliant platforms and GetLastError
+// on Windows.
+char* shared_lib_last_err();
+// As opposed to dlerror messages, those buffers will have to be
+// deallocated on Windows.
+void shared_lib_free_err_msg(char* msg);
+void shared_lib_print_last_err(std::ostream *ss);
+void* find_symbol(lib_handle handle, const char* symbol);
```
Comment (wjwithagen) [inline src/common/shared_lib.h]: Please make the naming consistent
And most of the time in ceph that would be `shared_lib_<function>`, so you'd need to fix close and open.

File: src/common/shared_lib.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#else
+#define SHARED_LIB_SUFFIX ".so"
+#endif
+
+#ifdef _WIN32
+  #include <windows.h>
+  typedef HMODULE lib_handle;
+#else
+  #include <ostream>
+  typedef void* lib_handle;
```
Comment (wjwithagen) [inline src/common/shared_lib.h]: Typedefs are usually looking like `some_type_t`
And in this case I'd call it `shared_lib_handle_t`

File: src/crimson/osd/CMakeLists.txt [hunk 3]
```diff
@@ -33,7 +33,7 @@ add_executable(crimson-osd
   watch.cc
   )
 target_link_libraries(crimson-osd
-  crimson-common crimson-os crimson fmt::fmt)
+  ceph-common crimson-common crimson-os crimson fmt::fmt)
 set_target_properties(crimson-osd PROPERTIES
```
Comment (wjwithagen) [inline src/crimson/osd/CMakeLists.txt]: Looks like this addition does not have to do anything with shared libs.
Unless it is the way to add `shared_lib` code.
if not you might want te remove it here, or at least make it a separate commit..

File: src/osd/CMakeLists.txt [hunk 4]
```diff
@@ -46,7 +46,7 @@ endif()
 add_library(osd STATIC ${osd_srcs})
 target_link_libraries(osd
   PUBLIC dmclock::dmclock
-  PRIVATE os heap_profiler cpu_profiler ${CMAKE_DL_LIBS})
+  PRIVATE os heap_profiler cpu_profiler ceph-common ${CMAKE_DL_LIBS})
```
Comment (wjwithagen) [inline src/osd/CMakeLists.txt]: Same here for the `ceph-common`

File: src/common/shared_lib.h (hunk 1 above)
Comment (petrutlucian94) [reply src/common/shared_lib.h]: Makes sense.

File: src/crimson/osd/CMakeLists.txt (hunk 3 above)
Comment (petrutlucian94) [reply src/crimson/osd/CMakeLists.txt]: ``shared_lib`` is part of ``ceph-common``, for which reason I had to link ``ceph-osd`` against it, otherwise it won't be able to use those functions.

File: src/osd/CMakeLists.txt (hunk 4 above)
Comment (petrutlucian94) [reply src/osd/CMakeLists.txt]: It's about the same issue, but it mainly affects Windows, which doesn't allow unresolved dependencies.

File: src/common/shared_lib.h (hunk 2 above)
Comment (petrutlucian94) [reply src/common/shared_lib.h]: ``shared_lib_handle_t`` sounds good.

**Comment by wjwithagen** [comment]
@petrutlucian94 
If this builds, I'm oke with it.

File: src/common/shared_lib.cc [hunk 5]
```diff
@@ -0,0 +1,74 @@
+#include "shared_lib.h"
+
+
+#if defined(_WIN32)
```
Comment (tchaikov) [inline src/common/shared_lib.cc]: @petrutlucian94 instead of introducing a set of functions for manipulating shared libraries, have you considered adding an other approach? something like: 

1. add a `dlfcn_win32.h` and `dlfcn_win32.cc`  for Win32. 
1. add a library named `win32::dl` for Win32, so and make sure instead of `-ldl` on Win32 this library is linked.

File: src/crimson/osd/CMakeLists.txt [hunk 6]
```diff
@@ -33,7 +33,7 @@ add_executable(crimson-osd
   watch.cc
   )
 target_link_libraries(crimson-osd
-  crimson-common crimson-os crimson fmt::fmt)
+  ceph-common crimson-common crimson-os crimson fmt::fmt)
```
Comment (tchaikov) [inline src/crimson/osd/CMakeLists.txt]: crimson is a project which tries to avoid linking against ceph-common. please don't do this.

File: src/erasure-code/ErasureCodePlugin.cc [hunk 7]
```diff
@@ -16,22 +16,18 @@
  */
 
 #include <errno.h>
-#include <dlfcn.h>
 
 #include "ceph_ver.h"
 #include "ErasureCodePlugin.h"
 #include "common/errno.h"
+#include "common/shared_lib.h"
```
Comment (tchaikov) [inline src/erasure-code/ErasureCodePlugin.cc]: we could have a `dlfcn_compat.h` which includes `<dlfcn.h>` on platforms which supports dlfcn functions, and include the header which introduces the homebrew shims otherwise.

File: src/common/shared_lib.cc (hunk 5 above)
Comment (petrutlucian94) [reply src/common/shared_lib.cc]: Yes, but we can't have a drop in replacement because the Windows API is slightly different (e.g. error message handling or dlopen flags) so it's probably better to have an abstraction on top rather than a lib trying to copy the semantics.

File: src/crimson/osd/CMakeLists.txt (hunk 6 above)
Comment (petrutlucian94) [reply src/crimson/osd/CMakeLists.txt]: I wish I could, but then we'll have unresolved symbols.

File: src/erasure-code/ErasureCodePlugin.cc (hunk 7 above)
Comment (petrutlucian94) [reply src/erasure-code/ErasureCodePlugin.cc]: That IMO can become even more confusing, as we'd be including ``dlfcn_compat.h`` but use ``shared_lib_*`` functions.

File: src/common/shared_lib.h [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+#ifndef SHARED_LIB_H_
+#define SHARED_LIB_H_
+
+#if !defined(_WIN32)
+  #include <dlfcn.h>
+#endif
+
+#ifdef __APPLE__
+#define SHARED_LIB_SUFFIX ".dylib"
```
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply src/common/shared_lib.h]: Yep, we can use that.

File: src/common/shared_lib.cc (hunk 5 above)
Comment (tchaikov) [reply src/common/shared_lib.cc]: @petrutlucian94 i assume your concern is `shared_lib_last_err()` and `shared_lib_free_err_msg()` pair. if that's the case we could have something like:
```c++
#ifdef WIN32
using last_error_t = std::string;
HMODULE dlopen(const char* filename, int /* flags */) {
  return  LoadLibrary(filename);
}

last_error_t dlerror() {
  DWORD err_code = ::GetLastError();
  LPSTR buffer = NULL;
  DWORD len = FormatMessageA(...);
  assert(len > 0);
  std::string s{buffer, len);
  LocalFree(buffer);
  return s;
}
#endif
```

regarding to "dlopen flags", IIUC, you just drop the flags in Windows implementation.

File: src/erasure-code/ErasureCodePlugin.cc (hunk 7 above)
Comment (tchaikov) [reply src/erasure-code/ErasureCodePlugin.cc]: sorry for the confusion, i assume that we go with the drop-in-replacement approach.

File: src/crimson/osd/CMakeLists.txt (hunk 6 above)
Comment (tchaikov) [reply src/crimson/osd/CMakeLists.txt]: crimson is based on the seastar library which does not support Windows. so if we go with the win32::dl library approach we can just avoid linking against it.

File: src/common/shared_lib.cc (hunk 5 above)
Comment (petrutlucian94) [reply src/common/shared_lib.cc]: You're right, that should work. I'll give it a try and come with a follow up.

File: src/erasure-code/ErasureCodePlugin.cc (hunk 7 above)
Comment (petrutlucian94) [reply src/erasure-code/ErasureCodePlugin.cc]: Oh, right. Makes sense then.

File: src/crimson/osd/CMakeLists.txt (hunk 6 above)
Comment (petrutlucian94) [reply src/crimson/osd/CMakeLists.txt]: Yes, you're right.

File: src/common/shared_lib.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply src/common/shared_lib.h]: Hmm, I tried adding the following to ``config-h.in.cmake`` but it came up as undefined in acconfig.h:

`#cmakedefine SHARED_LIB_SUFFIX "@CMAKE_SHARED_LIBRARY_SUFFIX@"`

I think I'll keep that explicit ``SHARED_LIB_SUFFIX`` definition for now.

[the remark this answers is not in the document]
Comment (tchaikov) [reply src/common/shared_lib.h]: we need to have
```c++
#cmakedefine CMAKE_SHARED_LIBRARY_SUFFIX "@CMAKE_SHARED_LIBRARY_SUFFIX@"
```
i can also come up with a follow-up change to work on this.

File: src/common/shared_lib.cc (hunk 5 above)
Comment (petrutlucian94) [reply src/common/shared_lib.cc]: I sent a follow up. Thanks a lot for this suggestion, it's much cleaner now.

True that dlerror now returns std::string instead of char* but that's a drop in replacement as well in this case.

File: src/common/shared_lib.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply src/common/shared_lib.h]: Thanks for the hint. I was hoping to avoid another define. I'll send a follow up.

File: src/common/CMakeLists.txt [hunk 9]
```diff
@@ -100,6 +100,11 @@ set(common_srcs
   util.cc
   version.cc)
 
+if(WIN32)
+  list(APPEND common_srcs
+    dlfcn_win32.cc)
+endif()
```
Comment (tchaikov) [inline src/common/CMakeLists.txt]: could you do something like
```cmake
if(WIN32)
  add_library(dlfcn_win32 STATIC dlfcn_win32.cc)
endif()
# ...
if(WIN32)
  list(APPEND ceph_common_deps dlfcn_win32)
endif()
```
?

File: src/common/dlfcn_win32.cc [hunk 10]
```diff
@@ -0,0 +1,57 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2020 SUSE LINUX GmbH
```
Comment (tchaikov) [inline src/common/dlfcn_win32.cc]: are you sure about this? in general, the copyright belongs to the entity which you work for.

File: src/common/CMakeLists.txt (hunk 9 above)
Comment (petrutlucian94) [reply src/common/CMakeLists.txt]: Sure. That makes sense, since there are cases in which you may not want to use ceph-common (e.g. crimson).

File: src/common/dlfcn_win32.cc (hunk 10 above)
Comment (petrutlucian94) [reply src/common/dlfcn_win32.cc]: Yes, Suse is sponsoring this work. Thanks for checking though.

File: src/common/dlfcn_win32.cc [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+  //FreeLibrary returns 0 on error, as opposed to dlclose.
+  return !FreeLibrary(handle);
+}
+
+void* dlsym(void* handle, const char* symbol) {
+  return (void*)GetProcAddress(handle, symbol);
+}
+
+dl_errmsg_t dlerror() {
+  // As opposed to dlerror messages, this has to be freed.
```
Comment (tchaikov) [inline src/common/dlfcn_win32.cc]: move this comment down before `LPSTR msg = NULL;`

File: src/common/dlfcn_win32.cc [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    FORMAT_MESSAGE_ALLOCATE_BUFFER |
+    FORMAT_MESSAGE_FROM_SYSTEM |
+    FORMAT_MESSAGE_IGNORE_INSERTS,
+    NULL,
+    err_code,
+    0,
+    (LPSTR) &msg,
+    0,
+    NULL);
+  if(!msgLen) {
```
Comment (tchaikov) [inline src/common/dlfcn_win32.cc]: please add a space after `if`.

File: src/include/dlfcn_compat.h [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  #define RTLD_NODELETE   0x01000
+
+  void* dlopen(const char *filename, int flags);
+  int dlclose(void* handle);
+  dl_errmsg_t dlerror();
+  void* dlsym(void* handle, const char* symbol);
+#else
+  #include <dlfcn.h>
+
+  typedef char* dl_errmsg_t;
```
Comment (tchaikov) [inline src/include/dlfcn_compat.h]: please be consistent when defining a type.
```suggestion
  using dl_errmsg_t = char*;
```

File: src/common/dlfcn_win32.cc [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+void* dlsym(void* handle, const char* symbol) {
+  return (void*)GetProcAddress(handle, symbol);
+}
+
+dl_errmsg_t dlerror() {
+  // As opposed to dlerror messages, this has to be freed.
+  DWORD err_code = ::GetLastError();
+  LPSTR msg = NULL;
+  size_t msgLen = FormatMessageA(
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/common/dlfcn_win32.cc]: this comment is not fully addressed. please note, we don't use camelCase.

File: src/common/dlfcn_win32.cc [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    NULL,
+    err_code,
+    0,
+    (LPSTR) &msg,
+    0,
+    NULL);
+  if (!msgLen) {
+    std::ostringstream msg_stream;
+    msg_stream << "Unknown error (" << err_code << ").";
+    return std::string(msg_stream.str());
```
Comment (tchaikov) [inline src/common/dlfcn_win32.cc:51]:
```suggestion
    return msg_stream.str();
```

as `ostringstream.str()` does return a `std::string`.

File: src/CMakeLists.txt [hunk 16]
```diff
@@ -409,6 +409,7 @@ endif()
 
 if(WIN32)
   list(APPEND ceph_common_deps ws2_32 mswsock)
+  list(APPEND ceph_common_deps dlfcn_win32)
```
Comment (tchaikov) [inline src/CMakeLists.txt:412]: could you revise the title of the commit message.

could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes. in this case, it could be "common:".

File: src/common/dlfcn_win32.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply src/common/dlfcn_win32.cc]: Right, I missed fixing the casing, sorry about that.

File: src/CMakeLists.txt (hunk 16 above)
Comment (petrutlucian94) [reply src/CMakeLists.txt:412]: Sure, I just wasn't aware that "common" is considered to be a subcomponent. I'll change the commit title right away.

File: src/include/buffer.h [hunk 17]
```diff
@@ -34,7 +34,7 @@
 #include <stdint.h>
 #include <string.h>
 
-#ifndef __CYGWIN__
+#if !defined(__CYGWIN__) && !defined(_WIN32)
```
Comment (tchaikov) [inline src/include/buffer.h:37]: could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes. we don't have a subcomponent dedicated to win32 at this moment. in this case, it could be "include/buffer.h". please see `git log src/include/buffer.h` for more examples.

the same applies to other commits in this PR.

Comment (petrutlucian94) [reply src/include/buffer.h]: Sure. FWIW, this was meant to be part of the next PR but I moved it by mistake.

I'd keep the win32 tag along with the subsystem names, it allows us to keep track of those patches much more easily.

Comment (petrutlucian94) [reply src/include/buffer.h]: There are about 60 commits in this chain, (some of the ones from  the windows.10 branch will be squashed though).

I'll fix their title as we move on and address other comments against the next PRs, I hope that's ok with you.

Comment (tchaikov) [reply src/include/buffer.h]: @petrutlucian94 could you address this comment. as the only commit in this PR still have "win32" in the prefix of its title. and i'd suggest remove it for the reason i explained just now.

**Comment by tchaikov** [comment]
@petrutlucian94 could you address https://github.com/ceph/ceph/pull/32704#discussion_r405301094?

File: src/include/buffer.h (hunk 17 above)
Comment (petrutlucian94) [reply src/include/buffer.h]: Would you be ok with labeling it in a different way so that we make the distinction between the "win32" tag and subsystems, such as [win32]? That's they way it showed up in the first PR, the one that got merged: https://github.com/ceph/ceph/pull/31981/commits

The reason why I insist on keeping the win32 tag is that it makes it easier to track those commits, which is crucial when having 60 of them. It also provides a scope so that anyone can tell right away that it's about Windows support.

Comment (tchaikov) [reply src/include/buffer.h]: @petrutlucian94 why do you want to tag a commit by adding win32 in the **prefix** of its title? what's the use case of this? can those who interested in finding the win32 related commits search the PR by "win32" label? or could you rename the title to something like 

> common: Add dlfcn_compat.h for win32

or anyone interested in this could just search win32 **in** the title?

Comment (tchaikov) [reply src/include/buffer.h]: furthermore, i am working on python3 migration, shall we add "python3" in the prefix of the commit title for SEO ? i doubt it. again, the title is not a collection of tags, it just summarizes the change.

Comment (petrutlucian94) [reply src/include/buffer.h]: That's one of the reasons. If it's clearly underlined that it's a tag, e.g. [win32], people are more likely to use it. Otherwise we might get "Add dlfcn_compat.h for Windows" or other similar forms, instead of the same tag.

It also keeps the commit messages shorted and avoids confusion. For example, I have a commit saying "[win32] Initialize WSA and TLS". Without the tag, most probably people wouldn't be able to tell right away what's it about, it's quite unlikely that someone would think that WSA stands for Windows Socket API.

Comment (tchaikov) [reply src/include/buffer.h]: then why not be more specific by putting "init WSA and TLS for win32" ?

Comment (petrutlucian94) [reply src/include/buffer.h]: > furthermore, i am working on python3 migration, shall we add "python3" in the prefix of the commit title for SEO ? i doubt it. again, the title is not a collection of tags, it just summarizes the change.

IMO, if it helps, can't see any reason why not. It's not just SEO, it's easier for people going through the commit title, without checking the description.

Comment (petrutlucian94) [reply src/include/buffer.h]: > for win32

I'll have 60 commits ending in with "for win32", that prefix tag seemed cleaner but I can go either way.

Comment (tchaikov) [reply src/include/buffer.h]: Lucian, i still think we should not do this in a free-style way. but i agree with you that the title should help people walking through the commits. another thing is that we group the pull requests by its component in our release notes, i don't think our script is smart enough to group the PRs by tags. probably one day, we will have more flexible and structured way to define a title, but i hope we can stick to the format of "<component>: <summary>" before that.

Comment (petrutlucian94) [reply src/include/buffer.h]: > another thing is that we group the pull requests by its component in our release notes

Makes sense, I'll change the title. I'm glad that we had this discussion though so that we don't have to get back to this topic.

**Comment by petrutlucian94** [comment]
jenkins test make check

File: src/osd/CMakeLists.txt (hunk 4 above)
Comment (tchaikov) [reply src/osd/CMakeLists.txt:49]: this breaks the build on bionic. 
```
dpkg-shlibdeps: error: cannot find library libceph-common.so.2 needed by debian/ceph-osd/usr/bin/ceph-osd (ELF format: 'elf64-x86-64' abi: '0201003e00000000'; RPATH: '')
```

see https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=bionic,DIST=bionic,MACHINE_SIZE=gigantic/39412//consoleFull

**Comment by petrutlucian94** [comment]
> this breaks the build on bionic.

That's odd, thanks for bringing it up. Anyway, I'm now using "win32_common" so that it won't affect linux builds.

For some reason github won't allow me to mark this conversation as resolved.

File: src/osd/CMakeLists.txt [hunk 18]
```diff
@@ -46,7 +46,7 @@ endif()
 add_library(osd STATIC ${osd_srcs})
 target_link_libraries(osd
   PUBLIC dmclock::dmclock
-  PRIVATE os heap_profiler cpu_profiler ${CMAKE_DL_LIBS})
+  PRIVATE os heap_profiler cpu_profiler win32_common ${CMAKE_DL_LIBS})
```
Comment (tchaikov) [inline src/osd/CMakeLists.txt:49]: where is `win32_common` target defined? this breaks the build.

Comment (petrutlucian94) [reply src/osd/CMakeLists.txt:49]: Oh, right. There was another patch from the chain that was adding it, my bad. Fixing it right away.

Comment (tchaikov) [reply src/osd/CMakeLists.txt:49]: so we are back to https://github.com/ceph/ceph/pull/32704#discussion_r407216945 .

Comment (petrutlucian94) [reply src/osd/CMakeLists.txt:49]: > so we are back to [#32704 (comment)](https://github.com/ceph/ceph/pull/32704#discussion_r407216945) .

What do you mean? win32_common is defined here [1] and doesn't affect linux builds:

File: src/CMakeLists.txt [hunk 19]
```diff
@@ -460,6 +461,10 @@ else()
     NAMELINK_SKIP)
 endif()
 
+if(WIN32)
+  set(win32_common ceph-common)
```
Comment (tchaikov) [inline src/CMakeLists.txt:465]: i'd avoid this.

Comment (petrutlucian94) [reply src/CMakeLists.txt:465]: That was @wjwithagen suggestion [1] and IMO it was a really good one. The idea is that as opposed to Linux shared libaries, Windows DLLs can't have unresolved symbols. It was much easier to do this rather than add checks wherever ceph-common was needed.

[1] https://github.com/ceph/ceph/pull/31981#discussion_r353360732

File: src/osd/CMakeLists.txt (hunk 18 above)
Comment (petrutlucian94) [reply src/osd/CMakeLists.txt:49]: > so we are back to [#32704 (comment)](https://github.com/ceph/ceph/pull/32704#discussion_r407216945) . this is just confusing as it seems osd now links against win32_common..
> 
> could you add a check, something like
> 
> ```cmake
> if(WIN32)
>   target_link_libraries(osd
>     PRIVATE ceph-common)
> fi()
> ```

That was @wjwithagen suggestion [1] and IMO it was a really good one. The idea is that as opposed to Linux shared libaries, Windows DLLs can't have unresolved symbols and libosd is in fact using ceph-common functions. It was much easier to do this rather than add checks wherever ceph-common was needed (quite a few places).

[1] #31981 (comment)

Comment (tchaikov) [reply src/osd/CMakeLists.txt:49]: > in fact using ceph-common functions

yeah, if you could help fix the FTBFS, that'd be better.

but i don't think your reply addresses my concern regarding to "osd links against win32_common".

> It was much easier to do this rather than add checks wherever ceph-common was needed (quite a few places).

in addition to "it is easier", i also want to make sure that it's correct and it looks correct.

Comment (petrutlucian94) [reply src/osd/CMakeLists.txt:49]: My concern is that I got two reviews, asking me to do opposite things. I'm ok either way as long as this PR merges. I can add a separate check and drop that "win32_common" variable, which is a convenient way of linking against ceph-common but only when targeting Windows.

Comment (tchaikov) [reply src/osd/CMakeLists.txt:49]: okay, let's wait for @wjwithagen 's input then. on the other hand,

> For now, we're mostly interested in the client side, allowing Windows hosts to consume rados, rbd and cephfs resources.

if what you are after is to build the client support, why would you want to build osd on windows? AFAICT, osd is one of the daemons, i.e., it's on server side.

Comment (petrutlucian94) [reply src/osd/CMakeLists.txt:49]: Unfortunately some of the dependencies cascaded to that. I'm doing a quick build so that I can say which one exactly.

FWIW, here's a subsequent commit that was using that win32_common variable, avoiding lots of IF(WIN32) checks: https://github.com/ceph/ceph/pull/32779/commits/c0736b703e5ccbba3bb9040816dff10178c10597

File: src/CMakeLists.txt (hunk 19 above)
Comment (wjwithagen) [reply src/CMakeLists.txt:465]: > 
> 
> That was @wjwithagen suggestion [1] and IMO it was a really good one. The idea is that as opposed to Linux shared libaries, Windows DLLs can't have unresolved symbols and libosd is in fact using ceph-common functions. It was much easier to do this rather than add checks wherever ceph-common was needed.
> 
> [1] [#31981 (comment)](https://github.com/ceph/ceph/pull/31981#discussion_r353360732)

@petrutlucian94 
I'm sorry about the mix of opinions. @tchaikov has to merge the PR in the end and he has final saying in this. I can only try to keep you going as much as I can.
But then I also have some discussion about my PRs over time.
And like we discussed, I really do not like the continous streams of `if(WIN32)` on a lot of places to add you windows lib.
IMHO we need something that is the CMake target `ceph-common`, which includes everything that is needed to link the common libraries. 
And perhaps it needs a different nameing..... Something like `ceph-common-extra`.
But that I'll leave up to @tchaikov to discuss with you

File: src/osd/CMakeLists.txt (hunk 18 above)
Comment (petrutlucian94) [reply src/osd/CMakeLists.txt]: Double checked, an luckily we don't actually need libosd. I may have mistaken it with libosdc or the dependency may have been removed. I've dropped this so I can mark this as resolved.

File: src/CMakeLists.txt (hunk 19 above)
Comment (petrutlucian94) [reply src/CMakeLists.txt]: Luckily we don't need libosd, so we can defer this discussion to the follow up PR. Renaming win32_common to something else sounds good as well.

Comment (tchaikov) [reply src/CMakeLists.txt]: @wjwithagen `if(WIN32)` is indeed a curse. i think we can use generator expression in cmake to introduce platform specific linkage of flags, something like
```cmake
target_link_libraries(
  some_library
  PUBLIC
    some_other_library
    $<$<PLATFORM_ID:Windows>:wsock32>
  PRIVATE
    $<$<PLATFORM_ID:Linux>:some_linux_library>
)
```

Comment (wjwithagen) [reply src/CMakeLists.txt]: @tchaikov 
It already looks a lot nicer.
But then still we are adding this set of GenExpr to every executable build that needs these extra libraries. So I guess it sort of depends on how many places it should.
And from memory, it was quite a few, since it triggered me to make a remark about it.

Having said all this, I read that for this PR it is no longer needed, so we should take this up, once @petrutlucian94 runs into it again.
And then deal with it.

Comment (petrutlucian94) [reply src/CMakeLists.txt]: > @wjwithagen `if(WIN32)` is indeed a curse. i think we can use generator expression in cmake to introduce platform specific linkage of flags, something like
> 
> ```cmake
> target_link_libraries(
>   some_library
>   PUBLIC
>     some_other_library
>     $<$<PLATFORM_ID:Windows>:wsock32>
>   PRIVATE
>     $<$<PLATFORM_ID:Linux>:some_linux_library>
> )
> ```

That seems fine as well, thanks for the suggestion.

Comment (petrutlucian94) [reply src/CMakeLists.txt]: @tchaikov I've updated the subsequent PR to use this syntax: https://github.com/ceph/ceph/pull/32779/commits/429fc2e0033dc527518145c775450a900218c77a#diff-edee2648008d3d163ca405b0c44c1cedR28

For what is worth, there were 16 occurrences.

**Comment by tchaikov** [comment]
```
[ RUN      ] TestMockMirrorDisableRequest.Success
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/mutex_debug.h: In function 'ceph::mutex_debug_detail::mutex_debug_impl<<anonymous> >::~mutex_debug_impl() [with bool Recursive = false]' thread 7f11c8a15700 time 2020-04-14T12:19:20.223929+0000
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/mutex_debug.h: 109: FAILED ceph_assert(r == 0)
 ceph version Development (no_version) pacific (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x19d) [0x7f11db68091b]
 2: (()+0x1599b9d) [0x7f11db680b9d]
 3: (ceph::mutex_debug_detail::mutex_debug_impl<false>::~mutex_debug_impl()+0x31) [0x55facedb044b]
 4: (()+0x1e615ce) [0x55facf53b5ce]
 5: (()+0x1e6476b) [0x55facf53e76b]
 6: (ContextWQ::process(Context*)+0xdd) [0x55facef61eff]
 7: (ThreadPool::PointerWQ<Context>::_void_process(void*, ThreadPool::TPHandle&)+0x32) [0x55facefc6a32]
 8: (ThreadPool::worker(ThreadPool::WorkThread*)+0x730) [0x7f11db6405f8]
 9: (ThreadPool::WorkThread::entry()+0x32) [0x7f11db644586]
 10: (Thread::entry_wrapper()+0x78) [0x7f11db61ddb4]
 11: (Thread::_entry_func(void*)+0x18) [0x7f11db61dd32]
 12: (()+0x76db) [0x7f11e4feb6db]
 13: (clone()+0x3f) [0x7f11d906388f]
```

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2020-04-14-2111/2e2cb50c291439df35a27c86948b101554716f2b/
